### ISSUE TITLE
Handling of ranges are inconsistent

### ISSUE THREAD

**Issue by crides** [nushell/nushell#11271]
### Describe the bug

Normal ranges are normal:
```
> 0..5
╭───┬───╮
│ 0 │ 0 │
│ 1 │ 1 │
│ 2 │ 2 │
│ 3 │ 3 │
│ 4 │ 4 │
│ 5 │ 5 │
╰───┴───╯
```
`range` command is also normal:
```
> 0..10 | each {$in} | range 0..5
╭───┬───╮
│ 0 │ 0 │
│ 1 │ 1 │
│ 2 │ 2 │
│ 3 │ 3 │
│ 4 │ 4 │
│ 5 │ 5 │
╰───┴───╯
```
`str substring` and `bytes at` are not:
```
> "012345" | str substring 0..5
01234

> "012345" | into binary | bytes at 0..5
Length: 5 (0x5) bytes | printable whitespace ascii_other non_ascii
00000000:   30 31 32 33  34                                      01234
```

### How to reproduce

See above

### Expected behavior

I understand that POSIX shells' ranges are inclusive, e.g. in zsh: 
```
> echo {0..5}
0 1 2 3 4 5
```
and slicing ranges that are open on the end is slighly easier to use (e.g. for calculating the resulting elements). That said, the behaviour between `range`, `str substring`, and `bytes at` should still be consistent. And, given that nushell does provide the "open at the end" ranges `$start..<$end` all 4 of these use cases should be consistent with each other.

TL;DR: `range`, `str substring` and `bytes at`, when given e.g. `0..5`, should all output max of 6 elements

> [!NOTE]
> There may be more cases where ranges are used, but these are the ones I've encountered so far. Please mention any other such commands/use cases, they should definitely be taken into account here

### Configuration

| key                | value                                     |
| ------------------ | ----------------------------------------- |
| version            | 0.86.1                                    |
| branch             | main                                      |
| commit_hash        | 8560b9b90e078974f908c0a49ea8fedd3ab2ded4  |
| build_os           | linux-x86_64                              |
| build_target       | x86_64-unknown-linux-gnu                  |
| rust_version       | rustc 1.71.1 (eb26296b5 2023-08-03)       |
| rust_channel       | 1.71.1-x86_64-unknown-linux-gnu           |
| cargo_version      | cargo 1.71.1 (7f1d04c00 2023-07-29)       |
| build_time         | 2023-11-14 23:19:33 -08:00                |
| build_rust_channel | release                                   |
| allocator          | mimalloc                                  |
| features           | default, extra, sqlite, trash, which, zip |
| installed_plugins  |                                           |

**Comment by fdncred** [comment]
What makes this worse is that it appears that we process ranges in multiple ways.
- There are commands like `echo` that I think use range parsing in the `nu_parser::parser::parse_range`
- There are commands that use `SyntaxShape::Range` and are/may be parsed with `cmd_base::util::process_range`
- There are commands that call for a range but take a `SyntaxShape::Any`
- maybe others?

It looks to me like c`md_base::util::process_range` probably needs to be aligned with `nu_parser::parser::parse_range`. The down side of this is that it will be a breaking change for several commands.

**Comment by TWSiO** [comment]
I think this is probably a duplicate of #7761, although with the additional identification of the same issue for `bytes at`.

**Comment by TWSiO** [comment]
Technically I suppose they are slightly different, but basically about the same issue and a solution to either would ideally solve both.

**Comment by fdncred** [comment]
agreed. close as a dupe of #7761
